src/blockdata/witness.rs

```diff
@@ -0,0 +1,327 @@
+                cursor += element_size_varint_len;
+                d.read_exact(&mut content[cursor..cursor + element_size])?;
+                cursor += element_size;
+            }
+            content.truncate(cursor);
+            Ok(Witness {
+                content,
+                witness_elements,
+            })
+        }
+    }
+}
+
+fn resize_if_needed(vec: &mut Vec<u8>, required_len: usize) {
+    if required_len >= vec.len() {
+        let mut new_len = vec.len().max(1);
+        while new_len <= required_len {
+            new_len *= 2;
+        }
+        vec.resize(new_len, 0);
+    }
+}
+
+impl Encodable for Witness {
+    fn consensus_encode<W: Write>(&self, mut writer: W) -> Result<usize, io::Error> {
+        let len = VarInt(self.witness_elements);
+        len.consensus_encode(&mut writer)?;
+        writer.emit_slice(&self.content[..])?;
+        Ok(self.content.len() + len.len())
+    }
+}
+
+impl Witness {
+    /// Returns `true` if the witness contains no element
+    pub fn is_empty(&self) -> bool {
+        self.witness_elements == 0
+    }
+
+    /// Returns a struct implementing [`Iterator`]
+    pub fn iter(&self) -> WitnessIterator {
+        WitnessIterator {
+            witness: &self,
+            cursor: 0,
+        }
+    }
+
+    /// Returns the number of elements this witness holds
+    pub fn len(&self) -> usize {
+        self.witness_elements as usize
+    }
+
+    /// Returns the bytes required when this Witness is consensus encoded
+    pub fn serialized_len(&self) -> usize {
+        self.iter()
+            .map(|el| VarInt(el.len() as u64).len() + el.len())
+            .sum::<usize>()
+            + VarInt(self.witness_elements).len()
+    }
+
+    /// Clear the witness
+    pub fn clear(&mut self) {
+        self.content.clear();
+        self.witness_elements = 0;
+    }
+
+    /// Push a new element on the witness, require an allocation
+    pub fn push(&mut self, new_element: &[u8]) {
+        self.witness_elements += 1;
+        let element_len_varint = VarInt(new_element.len() as u64);
+        let current_content_len = self.content.len();
+        self.content.resize(
+            current_content_len + element_len_varint.len() + new_element.len(),
+            0,
+        );
+        let end_varint = current_content_len + element_len_varint.len();
+        element_len_varint
+            .consensus_encode(&mut self.content[current_content_len..end_varint])
+            .expect("writers on vec don't error, space granted through previous resize");
+        self.content[end_varint..].copy_from_slice(new_element);
+    }
+}
+
+impl Default for Witness {
+    fn default() -> Self {
+        // from https://doc.rust-lang.org/std/vec/struct.Vec.html#method.new
+        // The vector will not allocate until elements are pushed onto it.
+        Witness {
+            content: Vec::new(),
+            witness_elements: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for WitnessIterator<'a> {
```

Comment:
Nit: I think the naming for iterator **structs** tends to end with `Iter`, not `Iterator` (e.g. `slice::Iter`).

Also see: https://rust-lang.github.io/api-guidelines/naming.html#iterator-type-names-match-the-methods-that-produce-them-c-iter-ty

Reply:
Yes, I believe so.